### PR TITLE
Update casted collision evaluator to handle fixed start and end states

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
This updates casted collision evaluator to handle fixed start or end states. Before we just would exclude but that is not ideal because you would periodically get post check failures  between the start and first waypoint and next to last waypoint and the end waypoint. 

Also this PR includes updating the constructors to pass eigen types by reference. The documentation states not to pass eigen types by value and I know in the past we have had eigen issues.

### CODE REVIEW

File: trajopt/src/collision_terms.cpp [hunk 1]
```diff
@@ -528,8 +876,43 @@ void CastCollisionEvaluator::CalcCollisions(const DblVec& x, tesseract_collision
...
+                  {
+                    return false;
+                  }
+                  case CollisionExpressionEvaluatorType::START_FIXED_END_FREE:
+                  {
+                    if (r.cc_type[0] == tesseract_collision::ContinuousCollisionType::CCType_Time0)
+                      return true;
+
+                    if (r.cc_type[1] == tesseract_collision::ContinuousCollisionType::CCType_Time0)
+                      return true;
```
Comment (mpowelson) [inline trajopt/src/collision_terms.cpp]: Should CCTYPE_NONE be removed?

Comment (mpowelson) [reply trajopt/src/collision_terms.cpp]: Also, do these enums follow the correct style?

File: trajopt/src/collision_terms.cpp [hunk 2]
```diff
@@ -421,6 +454,295 @@ void SingleTimestepCollisionEvaluator::Plot(const tesseract_visualization::Visua
...
+  double dist = (s1 - s0).norm();
+  long cnt = 2;
+  if (dist > longest_valid_segment_length_)
+  {
+    // Calculate the number state to interpolate
+    cnt = static_cast<long>(std::ceil(dist / longest_valid_segment_length_)) + 1;
+  }
+
+  // Get active link names
+  const std::vector<std::string>& active_links = adjacency_map_->getActiveLinkNames();
```
Comment (mpowelson) [inline trajopt/src/collision_terms.cpp]: This is not in the CastCollisionEvaluator. Should it be added around 812?

File: trajopt/src/collision_terms.cpp [hunk 3]
```diff
@@ -421,6 +454,295 @@ void SingleTimestepCollisionEvaluator::Plot(const tesseract_visualization::Visua
...
+        // If the contact pair does not exist in contact_results add it
+        if (p == contact_results.end() && !pair.second.empty())
+        {
+          contact_results[pair.first] = pair.second;
+        }
+        else
+        {
+          // Note: Must include all contacts through out the trajectory so the optimizer has all the information
+          //      to understand how to adjust the start and end state to move it out of collision. Originally tried
+          //      keeping the works case only but ran into edge cases where this does not work in the units tests.
```
Comment (mpowelson) [inline trajopt/src/collision_terms.cpp]: typos
`works case` -> `worst case`

`through out` -> `throughout`

File: trajopt/src/collision_terms.cpp (hunk 1 above)
Comment (Levi-Armstrong) [reply trajopt/src/collision_terms.cpp]: Are you suggesting removing it from the enum class?

Comment (mpowelson) [reply trajopt/src/collision_terms.cpp]: No, should it return true if it is that type.

Comment (Levi-Armstrong) [reply trajopt/src/collision_terms.cpp]: Right now CCType_None is to represent objects that are static in the world. You should never have a case where you checking two shapes that are of CCType_None which is why I only check for the CCType_Time0 or CCType_Time1 depending on the condition.

File: trajopt/src/collision_terms.cpp (hunk 2 above)
Comment (Levi-Armstrong) [reply trajopt/src/collision_terms.cpp]: This has been restructured.

**Comment by Levi-Armstrong** [comment]
@mpowelson I clean things up and added documentation to the collision evaluators. I also tested this on the examples and everything works as expected.

**Comment by mpowelson** [comment]
Remember to remove CI commit

File: trajopt/include/trajopt/collision_terms.hpp [hunk 4]
```diff
@@ -54,11 +100,57 @@ struct CollisionEvaluator
...
+
+  /**
+   * @brief Calculate the distance expressions when the end is free but the start is fixed
+   * @param x The current values
+   * @param exprs The returned expression
+   */
+  void CalcDistExpressionsEndFree(const DblVec& x, sco::AffExprVector& exprs);
+
+  /**
+   * @brief Calculate the distance expressions when the start and end is free
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply trajopt/include/trajopt/collision_terms.hpp]: fixed

**Comment by Levi-Armstrong** [comment]
I made the requested changes and remove the commit.
